### PR TITLE
[WIP][SPARK-49145] Improve readability of log4j console log output

### PR DESCRIPTION
**Pull request by asl3**
### What changes were proposed in this pull request?

This issue updates the default `log4j.properties.template` with the following improvements for console logging format:

- Use PatternLayout for improved human readability
- Color-code log levels (INFO/ERROR/WARN) to simplify logging output
- Visually partition the threadName and contextInfo for easy interpretation

See below for sample screenshots of the improved logs.

### Why are the changes needed?
Prior to this update, the OSS Spark logs were difficult to interpret. The logs followed a JSON output format which is not optimal for human consumption:

```

{"ts":"2024-07-26T18:45:17.712Z","level":"INFO","msg":"Running Spark version 4.0.0-SNAPSHOT","context":{"spark_version":"4.0.0-SNAPSHOT"},"logger":"SparkContext"}{"ts":"2024-07-26T18:45:17.715Z","level":"INFO","msg":"OS info Mac OS X, 14.4.1, aarch64","context":{"os_arch":"aarch64","os_name":"Mac OS X","os_version":"14.4.1"},"logger":"SparkContext"}{"ts":"2024-07-26T18:45:17.716Z","level":"INFO","msg":"Java version 17.0.11","context":{"java_version":"17.0.11"},"logger":"SparkContext"}{"ts":"2024-07-26T18:45:17.761Z","level":"WARN","msg":"Unable to load native-hadoop library for your platform... using builtin-java classes where applicable","logger":"NativeCodeLoader"}{"ts":"2024-07-26T18:45:17.783Z","level":"INFO","msg":"==============================================================","logger":"ResourceUtils"}{"ts":"2024-07-26T18:45:17.783Z","level":"INFO","msg":"No custom resources configured for spark.driver.","logger":"ResourceUtils"}{"ts":"2024-07-26T18:45:17.783Z","level":"INFO","msg":"==============================================================","logger":"ResourceUtils"}{"ts":"2024-07-26T18:45:17.784Z","level":"INFO","msg":"Submitted application: Spark Pi","context":{"app_name":"Spark Pi"},"logger":"SparkContext"}...{"ts":"2024-07-26T18:45:18.036Z","level":"INFO","msg":"Start Jetty 0.0.0.0:4040 for SparkUI","context":{"host":"0.0.0.0","port":"4040","server_name":"SparkUI"},"logger":"JettyUtils"}{"ts":"2024-07-26T18:45:18.044Z","level":"INFO","msg":"jetty-11.0.20; built: 2024-01-29T21:04:22.394Z; git: 922f8dc188f7011e60d0361de585fd4ac4d63064; jvm 17.0.11+9-LTS","logger":"Server"}{"ts":"2024-07-26T18:45:18.054Z","level":"INFO","msg":"Started Server@22c75c01{STARTING}[11.0.20,sto=30000] @1114ms","logger":"Server"} 

```

### Does this PR introduce _any_ user-facing change?
Yes, please see below for the updated log output. Note that users can override the provided `log4j.properties.template` as desired.

Screenshot 2024-08-07 at 10 38 25 AM

Screenshot 2024-08-07 at 10 38 55 AM

### How was this patch tested?
Producing console logs via spark-submit, e.g:

`./bin/spark-submit --class org.apache.spark.examples.SparkPi  examples/target/scala-2.13/jars/spark-examples_2.13-4.0.0-SNAPSHOT.jar 100`

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: conf/log4j2.properties.template
Comment (gengliangwang) [inline conf/log4j2.properties.template:1]: Let's keep another copy of json template

File: conf/log4j2.properties.template [hunk 1]
```diff
@@ -22,8 +22,8 @@ rootLogger.appenderRef.stdout.ref = console
 appender.console.type = Console
 appender.console.name = console
 appender.console.target = SYSTEM_ERR
-appender.console.layout.type = JsonTemplateLayout
-appender.console.layout.eventTemplateUri = classpath:org/apache/spark/SparkLayout.json
+appender.console.layout.type = PatternLayout
+appender.console.layout.pattern = %d{ISO8601} [%t] %c{1} - %highlight{%m} %notEmpty{%X}%n%n
```
Comment (gengliangwang) [inline conf/log4j2.properties.template:26]: let's move the context before the message

**Comment by gengliangwang** [comment]
If you are going to change the default, we should change log4j2-defaults.properties as well.

**Review by dongjoon-hyun — COMMENTED** [review]
Have you try `spark.log.structuredLogging.enabled=false`? I'm wondering if you are reporting `spark.log.structuredLogging.enabled` is not working correctly, @asl3 .

https://github.com/apache/spark/blob/5481d39177bbbffb5fe5ae65617ac6e06074ceb0/docs/configuration.md?plain=1#L3729-L3731

https://github.com/apache/spark/blob/5481d39177bbbffb5fe5ae65617ac6e06074ceb0/docs/core-migration-guide.md?plain=1#L47-L48

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
I guess we need another vote explicitly to revert the voted behavior. WDYT, @gengliangwang ?

- https://lists.apache.org/thread/z12gzntqrk5k0rt5h9voynsrgx2qdbyr
> [VOTE] SPIP: Structured Logging Framework for Apache Spark

**Comment by gengliangwang** [comment]
Yes, we definitely need a vote. I will start a discussion and vote after this PR is ready.

@asl3 could you mark this PR as DRAFT?

**Comment by dongjoon-hyun** [comment]
In addition, the migration guide and all related documents should be reverted accordingly. Please make it sure them too, @gengliangwang .

**Comment by asl3** [comment]
closing this PR so we can receive more feedback from the community

**Comment by asl3** [comment]
The PR is a prototype not yet ready for review. We will close it until we develop a comprehensive SPIP, which we will send for a vote to the community.

**Comment by dongjoon-hyun** [comment]
BTW, @gengliangwang and @asl3 , I don't think this requires SPIP-level effort. Since this is a revert to the original behavior, a `[DISCUSS]` thread and the official `[VOTE]` thread might be good enough.

**Pull request closed without merging**